### DISCUSSION TITLE
[@rollup/plugin-commonjs] How-to use with native node modules

### DISCUSSION THREAD

**Discussion by raphaelmenges** [rollup/plugins#1416] [Q&A]
Hello 👋

I am using Rollup through Vite 4 and have a hard time to bundle native Node.js modules that are (still) using commonjs style. In specifically, i am facing following issues:

1. Rollup [tries to parse](https://github.com/napi-rs/napi-rs/discussions/1361) required `.node` files, even though the [documentation](https://github.com/rollup/plugins/tree/master/packages/commonjs#include) states only `".cjs"` are considered by default.

```
[commonjs--resolver] Unexpected character '�' (Note that you need plugins to import files that are not JavaScript)
file: /Users/<app>/node_modules/<module>/index.js:1:0
```

2. Relative require paths to `.node` binaries seem to be resolved in relation to the project root, not the module that is requiring them.

```
Could not dynamically require "<project>\build\addon.node". Please configure the dynamicRequireTargets or/and ignoreDynamicRequires option of @rollup/plugin-commonjs appropriately for this require call to work.
```

I assume it might be fixable through choosing the correct [options](https://github.com/rollup/plugins/tree/master/packages/commonjs#options). Does anyone know details or has an example at hand how to handle modules in commonjs style that require `.node` binaries?

**Comment by koonopek** [comment]
@raphaelmenges  Have you managed to sole the issue?

**Reply by raphaelmenges** [reply]
I am using [this template](https://github.com/cawa-93/vite-electron-builder/blob/main/packages/main/vite.config.js) that handles everything as CJS. One can add ESM dependencies via 

```
ssr: {
    noExternal: ["<module-name>"],
  },
```

I solved this issue a few month ago and cannot remember 100% what lead me to this solution. Soon, Electron will support ESM natively. Then I need to investigate again.

**Comment by bobogogo1990** [comment]
```js
import { createRequire } from "module";
export const nodeRequire = createRequire(import.meta.url);
```

Please use `nodeRequire` to `require` your native module

**Reply by kyr0** [reply]
Works using `nodeRequire`, but I'm not happy xD This should just be an auto-optimization.
